### PR TITLE
Victoriametrics

### PR DESCRIPTION
**Pull request by hagen1778**

### CODE REVIEW

File: cmd/tsbs_load_victoriametrics/main.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	loader = load.GetBenchmarkRunner()
+	var urls string
+	flag.StringVar(&urls, "urls", "http://localhost:8428/write",
+		"Comma-separated list of VictoriaMetrics ingestion URLs(single-node or VMInsert)")
+	flag.Parse()
+
+	if len(urls) == 0 {
+		log.Fatalf("missing `urls` flag")
+	}
+	vmURLs = strings.Split(urls, ",")
```
Comment (tenmozes) [inline cmd/tsbs_load_victoriametrics/main.go:38]: should we add validation for urls here?

Comment (hagen1778) [reply cmd/tsbs_load_victoriametrics/main.go:38]: Don't think so. Wrong values will be immediately reported on attempt to create http request.

File: README.md [hunk 2]
```diff
@@ -13,6 +13,7 @@ Current databases supported:
 + ClickHouse [(supplemental docs)](docs/clickhouse.md)
 + CrateDB [(supplemental docs)](docs/cratedb.md)
 + SiriDB [(supplemental docs)](docs/siridb.md)
++ VictoriaMetrics [(supplemental docs)](docs/siridb.md)
```
Comment (valyala) [inline README.md]: `docs/siridb.md` must be replaced by `docs/victoriametrics.md`

File: cmd/tsbs_generate_data/main.go [hunk 3]
```diff
@@ -6,6 +6,7 @@
 // InfluxDB bulk load format
 // MongoDB BSON format
 // TimescaleDB pseudo-CSV format (the same as for ClickHouse)
+// VictoriaMetrics bulk load format (the same as for InfluxDB)
```
Comment (valyala) [inline cmd/tsbs_generate_data/main.go:9]: IMHO, Prometheus `remote_write` format must be added in the future. At the moment it is OK to have only InfluxDB bulk load format for VictoriaMetrics.

File: cmd/tsbs_generate_queries/databases/victoriametrics/common.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/uses/devops"
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/utils"
+	iutils "github.com/timescale/tsbs/internal/utils"
+	"github.com/timescale/tsbs/query"
+)
+
+type BaseGenerator struct{}
+
+// GenerateEmptyQuery returns an empty
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/common.go]: `returns an empty ...`?

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/uses/devops"
+	"github.com/timescale/tsbs/query"
+)
+
+// Devops produces PromQL queries for all the devops query types.
+type Devops struct {
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go:13]: Should `Devops` be private?

File: cmd/tsbs_generate_queries/databases/victoriametrics/common.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	// label to describe type of query
+	label string
+	// desc to describe type of query
+	desc string
+	// time range for query executing
+	interval *iutils.TimeInterval
+	// time period to group by in seconds
+	step string
+	// whether to use instant query
+	instant bool
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/common.go]: Let's remove instant queries for now, since they don't return last values for the selected time series

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+func (d *Devops) GroupByOrderByLimit(qi query.Query) {
+	panic("GroupByOrderByLimit not supported in PromQL")
+}
+
+// LastPointPerHost finds the last row for every host in the dataset
+// e.g. in pseudo-PromQL:
+// sum({__name__=~"cpu_.*"}) by(hostname)
+func (d *Devops) LastPointPerHost(qq query.Query) {
+	qi := &queryInfo{
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: This doesn't return the last point - it returns only last points for the last 5 minutes. Remove this implementation

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+// 		{__name__=~"metric1|metric2...|metricN",hostname=~"hostname1|hostname2...|hostnameN"}[1m]
+// 	)
+// ) by (__name__)
+func (d *Devops) GroupByTime(qq query.Query, nHosts, numMetrics int, timeRange time.Duration) {
+	metrics := devops.MustGetCPUMetricsSlice(numMetrics)
+	hosts := d.MustGetRandomHosts(nHosts)
+	selectClause := getSelectClause(metrics, hosts)
+	qi := &queryInfo{
+		query:    fmt.Sprintf("max(max_over_time(%s)) by (__name__)", selectClause),
+		label:    fmt.Sprintf("VictoriaMetrics %d cpu metric(s), random %4d hosts, random %s by 1m", numMetrics, nHosts, timeRange),
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go:40]: `%4d`?

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+// 	max_over_time(
+// 		{__name__=~"metric1|metric2...|metricN",hostname=~"hostname1|hostname2...|hostnameN"}[1m]
+// 	)
+// ) by (__name__)
+func (d *Devops) GroupByTime(qq query.Query, nHosts, numMetrics int, timeRange time.Duration) {
+	metrics := devops.MustGetCPUMetricsSlice(numMetrics)
+	hosts := d.MustGetRandomHosts(nHosts)
+	selectClause := getSelectClause(metrics, hosts)
+	qi := &queryInfo{
+		query:    fmt.Sprintf("max(max_over_time(%s)) by (__name__)", selectClause),
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: forgot adding `[1m]` inside `max_over_time()`?

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// GroupByTimeAndPrimaryTag selects the AVG of numMetrics metrics under 'cpu' per device per hour for a day,
+// e.g. in pseudo-PromQL:
+//
+// avg(
+// 	avg_over_time(
+// 		{__name__=~"metric1|metric2...|metricN"}[1h]
+// 	)
+// ) by (hostname)
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: `by (__name__, hostname)`

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// MaxAllCPU selects the MAX of all metrics under 'cpu' per hour for nhosts hosts,
+// e.g. in pseudo-PromQL:
+//
+// max(
+// 	max_over_time(
+// 		{hostname=~"hostname1|hostname2...|hostnameN"}[1h]
+// 	)
+// )
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: `by (__name__)`

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	d.fillInQuery(qq, qi)
+}
+
+// HighCPUForHosts populates a query that gets CPU metrics when the CPU has high
+// usage between a time period for a number of hosts (if 0, it will search all hosts),
+// e.g. in pseudo-PromQL:
+//
+// max(
+// 	{__name__=~'cpu_.*', hostname=~"hostname1|hostname2...|hostnameN"}
+// ) by (cpu_usage_user) > 90
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: `by (hostname)`

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+func (d *Devops) HighCPUForHosts(qq query.Query, nHosts int) {
+	var hostClause string
+	if nHosts > 0 {
+		hosts := d.MustGetRandomHosts(nHosts)
+		hostClause = getHostClause(hosts)
+	}
+	qi := &queryInfo{
+		query:    fmt.Sprintf("max(cpu_usage_user{%s}) > 90", hostClause),
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: missing `by (hostname)`?

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+		label:    devops.GetMaxAllLabel("VictoriaMetrics", nHosts),
+		interval: d.Interval.MustRandWindow(devops.HighCPUDuration),
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go:110]: `interval` should be short in order to have only a single point. Now it returns two points and for the first point it requires scanning additional 12 hours of data points.

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go (hunk 13 above)
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: missing `max_over_time`?

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	d.fillInQuery(qq, qi)
+}
+
+// HighCPUForHosts populates a query that gets CPU metrics when the CPU has high
+// usage between a time period for a number of hosts (if 0, it will search all hosts),
+// e.g. in pseudo-PromQL:
+//
+// max(
+// 	{__name__=~'cpu_.*', hostname=~"hostname1|hostname2...|hostnameN"}
```
Comment (valyala) [inline cmd/tsbs_generate_queries/databases/victoriametrics/devops.go]: missing `max_over_time`?

File: cmd/tsbs_load_victoriametrics/processor.go [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (p *processor) do(b *batch) (uint64, uint64) {
+	for {
+		r := bytes.NewReader(b.buf.Bytes())
+		req, err := http.NewRequest("POST", p.url, r)
+		if err != nil {
+			log.Fatalf("error while creating new request: %s", err)
+		}
+		req.Header.Add("Content-Encoding", "snappy")
```
Comment (valyala) [inline cmd/tsbs_load_victoriametrics/processor.go]: Influx line protocol supports only `gzip` compression.

File: cmd/tsbs_load_victoriametrics/processor.go [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+		req.Header.Set("Content-Type", "application/x-protobuf")
```
Comment (valyala) [inline cmd/tsbs_load_victoriametrics/processor.go]: This header isn't needed

File: cmd/tsbs_run_queries_victoriametrics/main.go [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"strings"
+	"time"
+
+	"bytes"
+	_ "github.com/kshvakov/clickhouse"
```
Comment (valyala) [inline cmd/tsbs_run_queries_victoriametrics/main.go]: I think this import isn't needed

File: docs/victoriametrics.md [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+just a single-version URL or list of VMInsert URLs. Workers will be
+distributed in a round robin fashion across the URLs.
+
+---
+
+## Generating queries
+
+VictoriaMetrics supports PromQL which has some limitations
+comparing to SQL. Because of this VcitoriaMetrics query generator
+lacks of query type `groupby-orderby-limit` implementation.
```
Comment (valyala) [inline docs/victoriametrics.md]: `last-point` query is also missing

File: internal/inputs/generator_data.go [hunk 20]
```diff
@@ -227,6 +227,8 @@ func (g *DataGenerator) getSerializer(sim common.Simulator, format string) (seri
 	switch format {
 	case FormatCassandra:
 		ret = &serialize.CassandraSerializer{}
+	case FormatVictoriaMetrics:
+		fallthrough
```
Comment (valyala) [inline internal/inputs/generator_data.go]: it would be better from readability and maintainability PoV writing `ret = &serialize.InfluxSerializer{}` here instead of `fallthrough`

File: internal/inputs/generator_queries.go [hunk 21]
```diff
@@ -195,67 +196,81 @@ func (g *QueryGenerator) init(config GeneratorConfig) error {
 	return nil
 }
 
+type factory struct {
+	database  string
+	generator interface{}
+}
+
 func (g *QueryGenerator) initFactories() error {
-	cassandra := &cassandra.BaseGenerator{}
```
Comment (valyala) [inline internal/inputs/generator_queries.go]: This refactoring must be sent in a separate pull request to upstream repository, since it is orthogonal to the rest of changes.

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go (hunk 5 above)
Comment (hagen1778) [reply cmd/tsbs_generate_queries/databases/victoriametrics/devops.go:13]: Nope, implemented similar to other databases

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go (hunk 14 above)
Comment (hagen1778) [reply cmd/tsbs_generate_queries/databases/victoriametrics/devops.go:110]: What do you suggest?

File: cmd/tsbs_load_victoriametrics/processor.go (hunk 16 above)
Comment (hagen1778) [reply cmd/tsbs_load_victoriametrics/processor.go]: What if delete headers setting at all?

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go (hunk 8 above)
Comment (hagen1778) [reply cmd/tsbs_generate_queries/databases/victoriametrics/devops.go:40]: this is how every other generator formats label for this type of query

**Comment by albertdb** [comment]
When generating queries, I get the following error:

`error: use case 'iot' not implemented for format 'victoriametrics'`

Please include at least a warning about this in docs, not only about query types `groupby-orderby-limit` and `lastpoint`

File: cmd/tsbs_generate_queries/databases/victoriametrics/devops.go (hunk 14 above)
Comment (valyala) [reply cmd/tsbs_generate_queries/databases/victoriametrics/devops.go:110]: Probably sending the request to `/api/v1/query` with `time` arg set to the end of the interval.

**Review by valyala — APPROVED** [review]
